Make resource independent of data bags

frank-m opened the pull request:
# Description

This pull request removes the hard depedency on data bags from the manage resource. As such the user can choose if they want to use attributes/data bags or a different solution of their preference.

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

ramereth commented:
@frank-m are you intending on completely removing support of databags or only making it optional?

frank-m commented:
Make it optional. You can still pass the content of a data bag as the users property, but in the recipe instead of the the resource. I will provide an upgrading.md with details and make sure there is tests for that.

frank-m commented:
So, I updated the unit and integration tests and made sure that both data bag and non data bag usage is tested. I also updated the readme and added upgrading.md.

Can someone do a first review? After that I will remove the WIP: and squash the commits.

ramereth commented on README.md:
```diff
@@ -10,14 +10,16 @@ Manages OS users from databags.
 ## Scope
 
-This cookbook is concerned with the management of OS users and groups from databags. It also manages the distribution of ssh keys to a user's home directory.
+This cookbook is concerned with the management of OS users and groups. It also manages the distribution of ssh keys to a user's home directory.
```
I think this might work better.
```suggestion
This cookbook is concerned with the management of OS users and groups (optionally from databags). It also manages the distribution of ssh keys to a user's home directory.
```

ramereth commented on README.md:
```diff
@@ -10,14 +10,16 @@ Manages OS users from databags.
 
 ## Requirements
 
+If you are upgrading from a version < 5.7.0 please see [upgrading.md](https://github.com/sous-chefs/users/upgrading.md)
```
```suggestion
If you are upgrading from a version < 6.0.0 please see [upgrading.md](https://github.com/sous-chefs/users/upgrading.md)
```

ramereth commented on README.md:
```diff
@@ -58,7 +60,7 @@ Then in a recipe use the `user_manage` resource to add all users in the defined
   group_id GROUPID
   action [:create]
-  data_bag 'DATABAG_NAME'
+  users users_variable
```
Ideally, it'd be nice to still show examples for both with and without databags to help users.

frank-m replied:
I am showing three examples. One where I use a data bag, one where I use attributes and one where I use a local variable. So I think it is covered. If not, can you let me know what exampel you want specifically?

ramereth replied:
Ah yes, I should have looked at the whole file as rendered. Something I think would be helpful here is to show what `users_variable` should look like or just show a simple example without using a variable.

ramereth commented on README.md:
```diff
@@ -126,80 +162,60 @@ Other potential fields (optional):
 ### users_manage
 
-The `users_manage` resource manages users and groups based off of a data bag search and specified action.
+The `users_manage` resource manages users and groups based off the `users` property and the specified action(s).
```
```suggestion
The `users_manage` resource manages users and groups based off the `users` property or of a data bag search and the specified action(s).
```

ramereth commented on upgrading.md:
```diff
@@ -0,0 +1,35 @@
+
+## Major Changes
+
+Data bags are no longer directly supported. Before v6.0.0 you would do the following:
```
Why does this change have to be made? Can't we keep this functionality in the resource while also having the new functionality? This is a *really* big change for this cookbook and will impact the users quite a bit, even with a major version bump.

frank-m replied:
I discussed this briefly in the sous-chefs channel and the consensus was that it was better to make it independent of the data source. 

I think adding one line of code to load the data bag is not too bad for a major version upgrade right?

If you disagree I can refactor it to make a source or a data bag possible. But I personally do think this way is cleaner and it does less chef searches as well if you have multiple invocations of the resource.

ramereth replied:
Ah yes, I didn't get a chance to read the backlog. That makes sense.

bmhughes commented on resources/manage.rb:
```diff
@@ -17,79 +17,82 @@
+  # Break the loop if users is no in the specified group_name
+  new_resource.users.each do |user|
+    next unless user[:groups].include? new_resource.group_name
+    next if user[:action] == 'remove'
```
Leave a newline after guards

bmhughes commented on resources/manage.rb:
```diff
@@ -99,55 +102,55 @@
-      if u['ssh_private_key']
-        key_type = u['ssh_private_key'].include?('BEGIN RSA PRIVATE KEY') ? 'rsa' : 'dsa'
+      if user[:ssh_private_key]
+        key_type = user[:ssh_private_key].include?('BEGIN RSA PRIVATE KEY') ? 'rsa' : 'dsa'
```
I'd move this into a helper method so it'll be easier to add support for ECDSA and ED25519 etc

bmhughes commented on resources/manage.rb:
```diff
@@ -99,55 +102,55 @@
-      if u['ssh_public_key']
-        key_type = u['ssh_public_key'].include?('ssh-rsa') ? 'rsa' : 'dsa'
+      if user[:ssh_public_key]
+        key_type = user[:ssh_public_key].include?('ssh-rsa') ? 'rsa' : 'dsa'
```
Same as above

bmhughes commented on resources/manage.rb:
```diff
@@ -17,79 +17,82 @@
     # Set home to location in data bag,
     # or a reasonable default ($home_basedir/$user).
-    home_dir = (u['home'] || "#{home_basedir}/#{u['username']}")
+    home_dir = (user[:home] || "#{home_basedir}/#{username}")
```
Braces aren't needed

bmhughes reviewed (changes requested):
Few ruby syntax changes for the manage resource. If you run rubocop/cookstyle there's probably a few more.

ramereth commented on README.md:
```diff
@@ -10,14 +10,16 @@ Manages OS users from databags.
 ## Scope
 
-This cookbook is concerned with the management of OS users and groups from databags. It also manages the distribution of ssh keys to a user's home directory.
+This cookbook is concerned with the management of OS users and groups (optionally from databags). It also manages the distribution of ssh keys to a user's home directory.
```
You should also update the description on line 9.

frank-m commented on resources/manage.rb:
```diff
@@ -17,79 +17,82 @@
       ssh_keys = []
-      if u['ssh_keys']
-        Array(u['ssh_keys']).each do |key|
+      if user[:ssh_keys]
```
They can also provide a string. So, I suppose it would break then if Array() is not there right?

bmhughes replied:
Meant the if/end statement, not the Array() call, that covers you against nils.

bmhughes commented on resources/manage.rb:
```diff
@@ -17,79 +17,82 @@
-# :group_name is the string name of the group to create, defaults to resource name
-# :group_id is the numeric id of the group to create, default is to allow the OS to pick next
-# :cookbook is the name of the cookbook that the authorized_keys template should be found in
-property :data_bag, String, default: 'users'
```
https://docs.chef.io/release_notes_client/#deprecations-in-cookbook-resources

ramereth replied:
Actually, you want `deprecated`:
```
property :thing2, String, deprecated: 'The thing2 property has been deprecated and will be removed in the next major release of this cookbook scheduled for 12/25/2018!'
```

bmhughes replied:
Yes indeed, didn't catch the data_bag bit first time

frank-m replied:
But If I add that, then I should leave the functionality in right? It can't be deprecated and already not working right? Or is that okay?

ramereth replied:
No you don't need to leave the functionality as it will do the following:
```
Starting Chef Infra Client, version 16.10.17
Patents: https://www.chef.io/patents
Creating a new client identity for default-centos-7 using the validator key.
resolving cookbooks for run list: ["users_test::default"]
Synchronizing Cookbooks:
  - users_test (1.7.1)
  - users (5.7.0)
Installing Cookbook Gems:
Compiling Cookbooks...
[2021-03-10T20:17:57+00:00] ERROR: Deprecated resource property used from /opt/kitchen/cache/cookbooks/users_test/recipes/default.rb:9:in `block in from_file'

  this has been deprecated!

Please consult the resource documentation for more information.

================================================================================
Recipe Compile Error in /opt/kitchen/cache/cookbooks/users_test/recipes/default.rb
================================================================================

Chef::Exceptions::DeprecatedFeatureError
----------------------------------------
Deprecated resource property used from /opt/kitchen/cache/cookbooks/users_test/recipes/default.rb:9:in `block in from_file'

  this has been deprecated!

Please consult the resource documentation for more information. (raising error due to treat_deprecation_warnings_as_errors being set)

Cookbook Trace: (most recent call first)
----------------------------------------
  /opt/kitchen/cache/cookbooks/users_test/recipes/default.rb:9:in `block in from_file'
  /opt/kitchen/cache/cookbooks/users_test/recipes/default.rb:6:in `from_file'
```

Also, it looks like we need to bump the required version of Chef to at least 13.9 or greater in `metadata.rb` since that feature wasn't included in older versions.

frank-m replied:
I added the deprecated message and bumped the metadata to 14 as that is the lowest support version and it is quite close to 13.9.

frank-m commented:
@ramereth @bmhughes I just ran into something while implementing the users cookbook that I want to add as a feature.

It allows you to specify if a user needs to have a group with its own name or not.

Do you think I can add it to this pull request so we can release it into 6.0.0, or do you want me to open a separate one later?

bmhughes commented:
No it can go into this as it's a major release anyway

bmhughes commented on libraries/helpers.rb:
```diff
@@ -74,5 +76,19 @@ def home_basedir
+    # Returns the type of the public key or nil
+    #
+    # @return [String]
+    def pubkey_type(pubkey)
```
This can be simplified to something like:
```ruby
def pubkey_type(pubkey)
  %w(ed25519 dss rsa dsa ecdsa).filter { |kt| kt.eql?(pubkey.split.first) }.first || 'rsa'
end
```
If nothing matches `filter` will return an empty array which `.first` on it will return `nil`, so it makes it a nice one-liner.

frank-m replied:
Almost worked the first time.

Had to reorder the array to make sure dsa and ecdsa both work and use include instead of eql :) But quite a nice one liner indeed.

bmhughes commented on libraries/helpers.rb:
```diff
@@ -57,14 +57,16 @@ def shell_is_valid?(shell_path)
+    # handles basic input validation and checking 
+    # whether uid was specified as a string
     def validate_id(id)
+      return if id.nil? || id == 0
```
Leave a newline after a guard. Also i'd rethink this method generally as there's a mixture of types going on and nothing to handle them.

bmhughes commented on libraries/helpers.rb:
```diff
@@ -57,14 +57,16 @@ def shell_is_valid?(shell_path)
+    # whether uid was specified as a string
     def validate_id(id)
+      return if id.nil? || id == 0
       id.to_i.to_s == id ? id.to_i : id
```
Also, I'm really not sure what this is supposed to achieve here, if we just need to make sure that id is an `integer` then you can just call `.to_i` on whatever is passed in.

bmhughes replied:
This should be for line 64

frank-m replied:
Yeah most of the things you are commenting on were already in the code. I don't mind refactoring it though, so I will have a look at it :) 

I mean we could just kill the whole validate_id and just use .to_i in the resource? I mean we are already validating if the hash has a key called 'gid' or 'uid'.  The worst thing that would happen is that nil would become 0, which would just error out as you won't be able to overwrite root.

What do you think?

frank-m replied:
I decided to go for it as it does not really make sense to have a method to check if something is an integer :) validate_id has been killed off.

bmhughes commented on libraries/helpers.rb:
```diff
@@ -74,5 +66,12 @@ def home_basedir
+    #
+    # @return [String]
+    def pubkey_type(pubkey)
+      %w(ed25519 ecdsa dss rsa dsa).filter { |kt| pubkey.split.first.include? kt }.first || 'rsa'
```
Just stick some braces around the `include?` argument and I think we can call this done for now 😀

frank-m replied:
At the risk of sounding a bit stupid. Do you mean curly braces around the `kt` in this case?

Like this:

```%w(ed25519 ecdsa dss rsa dsa).filter { |kt| pubkey.split.first.include? { kt } }.first || 'rsa'```

If my assumption is correct, what does that achieve? I never used include with the braces before and `kt` is already an interator right so it does not need to be executed before hand?

frank-m replied:
Aaah, I just checked your previous example and I suppose you mean parantheses here, so I will go with that for now. If you do mean curly braces then please let me know because I'd like to learn the why :)

bmhughes replied:
Yes I did mean parantheses, apologies there.

bmhughes commented on resources/manage.rb:
```diff
@@ -16,156 +16,165 @@
+          variables private_key: user[:ssh_private_key]
         end
       end
+
```
Extra newline

bmhughes commented on resources/manage.rb:
```diff
@@ -16,156 +16,165 @@
+  # Loop through all the users in the users_hash
+  # Break the loop if users is no in the specified group_name
+  new_resource.users.each do |user|
+    next unless user[:groups].include? new_resource.group_name
```
Braces again

frank-m commented:
@bmhughes did you also see the new functionality I added?

bmhughes commented:
The `username_group` stuff? Yes I did, the branch could do with a good squash/rebase/reword when you're finished as it's a little hard to follow what's been done in what commit.

Also, you're probably going to hate me for this, but looking at this as a whole from the outside (I don't personally use this cookbook) I'd consider refactorting it to operate on a single user per resource and leave the multiple-user stuff to the wrapping cookbook. If you did this all the stuff around variables types and checks go away as you aren't dependant on an arbitrary hash being passed in that you *hope* has the correct types and values in it. As it stands this is a recipe implemented as a resource, I might be missing the point by not being a user of the cookbook though.

frank-m commented:
I thought about that, but that would be a massive breaking change I reckon.

No problem to do that, but not sure if the users of the users cookbook would still want to use it then :)

ramereth commented:
> I thought about that, but that would be a massive breaking change I reckon.
> 
> No problem to do that, but not sure if the users of the users cookbook would still want to use it then :)

Yeah kind of the point of this cookbook is being able to manage a group of users easily. I don't think it'd be a good idea to make that change.

frank-m commented:
I think I have addressed everything. Can you approve if you agree?

kitchen-porter commented:
Released as: [6.0.0](https://github.com/sous-chefs/users/releases/tag/6.0.0)
